### PR TITLE
Add e-mail-based sign in challenge for users with disabled 2FA

### PR DESCRIPTION
**Pull request by Gargron**

### CODE REVIEW

**Comment by shleeable** [comment]
Cool

File: app/controllers/concerns/two_factor_authentication_concern.rb [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+  rescue OpenSSL::Cipher::CipherError
+    false
+  end
+
+  def authenticate_with_two_factor
+    user = self.resource = find_user
+
+    if user_params[:otp_attempt].present? && session[:attempt_user_id]
+      authenticate_with_two_factor_attempt(user)
+    elsif user.present? && (user.encrypted_password.blank? || user.valid_password?(user_params[:password]))
```
Comment (ThibG) [inline app/controllers/concerns/two_factor_authentication_concern.rb:26]: I feel like keeping the comments about getting the account from LDAP or PAM would make sense.

**Review by ClearlyClaire — CHANGES_REQUESTED** [review]
This PR changes the login process so that users who do not have 2FA enabled have to—after successfully entering their password—enter a code sent to their e-mail address if they are logging from a new IP address **and** have not logged in in the past 2 weeks.

Those conditions seem reasonable. We may want to require that challenge on new IP addresses without the second condition, but I'm afraid this could be too annoying, especially with mobile users or people using the Tor Browser. Then again, they could be using 2FA to avoid the mail confirmation.

In addition to the rather minor comment I made inline, tests are failing because they pass session data with `otp_user_id` that you renamed to `attempt_user_id`.

**Review by ClearlyClaire — APPROVED** [review]

**Comment by ThreeFx** [comment]
Is there a way to disable this option? Every Mastodon instance without sending E-Mail support runs into issues here.

**Comment by Gargron** [comment]
> Every Mastodon instance without sending E-Mail support

Running Mastodon without being able to send e-mails is impractical, so I would hope those are rare. If your users use 2FA to protect their accounts you won't run into the issue though.

**Comment by ThreeFx** [comment]
Alright I kind of expected this. Thanks for the quick answer!

**Pull request merged**